src/rulenode.jl

```diff
@@ -199,9 +191,12 @@ are found.
 Base.isless(rn₁::AbstractRuleNode, rn₂::AbstractRuleNode)::Bool = _rulenode_compare(rn₁, rn₂) == -1
 
 
-function _rulenode_compare(rn₁::RuleNode, rn₂::RuleNode)::Int
+function _rulenode_compare(rn₁::AbstractRuleNode, rn₂::AbstractRuleNode)::Int
     # Helper function for Base.isless
-    if rn₁.ind == rn₂.ind
+    if !isfilled(rn₁) || !isfilled(rn₂)
+        throw("BadArgument: unable to compare nodes of types ($(typeof(rn₁)), $(typeof(rn₂)))")
```

Comment:
This is probably something to wait to address now, but handling this with multiple dispatch would be nice. We would have to rework how `isfilled` works—maybe reflecting that info as separate types/with traits.

Ignore this for now. Just making note of it for the future.

Reply:
I'm not sure how to implement traits in julia. Is is possible to use julia's dispatch system to dispatch based on traits?

One thing to keep in mind is that `StateHole`s often switch between being filled and not being filled. So this property can toggle on and off for one hole instance.

Reply:
That's a good point. Like I said, don't worry about this for now, was more thinking out loud than anything.